File: notification-service/src/main/java/com/nixtap/notificationservice/service/NotificationService.java
package com.nixtap.notificationservice.service;

import com.nixtap.notificationservice.dto.request.NotificationPreferenceRequest;
import com.nixtap.notificationservice.dto.request.SendNotificationRequest;
import com.nixtap.notificationservice.dto.response.NotificationLogResponse;
import com.nixtap.notificationservice.dto.response.NotificationPreferenceResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface NotificationService {
    /** Get or auto-create notification preferences for the authenticated user. */
    NotificationPreferenceResponse getMyPreferences();

    /** Update preferences for the authenticated user. */
    NotificationPreferenceResponse updateMyPreferences(NotificationPreferenceRequest request);

    /** Paginated notification log for the authenticated user. */
    Page<NotificationLogResponse> getMyLogs(Pageable pageable);

    /** Count unread notifications for the authenticated user. */
    long countUnread();

    /** Mark a single log entry as read. */
    void markAsRead(Long logId);

    /** Mark all log entries as read for the authenticated user. */
    void markAllAsRead();

    /**
     * Internal: dispatch a notification respecting user preferences.
     * Called by other services via Feign — not exposed to end users.
     */
    void sendNotification(SendNotificationRequest request);
}

File: nixtap-frontend/src/api/notificationService.js
import api from './axios';

// ─── NOTIFICATION SERVICE  /api/v1/notifications ─────────────────────────

/** Get my notification preferences */
export const getNotificationPreferences = async () => {
  try {
    const res = await api.get('/api/v1/notifications/preferences');
    return res.data;
  } catch {
    return { data: null };
  }
};

/** Update my notification preferences */
export const updateNotificationPreferences = async (data) => {
  const res = await api.put('/api/v1/notifications/preferences', data);
  return res.data;
};

/** Get my notification logs (paginated) */
export const getNotificationLogs = async (page = 0, size = 20) => {
  try {
    const res = await api.get('/api/v1/notifications/logs', { params: { page, size } });
    return res.data;
  } catch {
    return { data: { content: [] } };
  }
};

/** Get count of unread notifications */
export const getUnreadNotificationCount = async () => {
  try {
    const res = await api.get('/api/v1/notifications/logs/unread/count');
    return res.data;
  } catch {
    return { data: 0 };
  }
};

/** Mark a single notification log as read */
export const markNotificationRead = async (id) => {
  const res = await api.put(`/api/v1/notifications/logs/${id}/read`);
  return res.data;
};

/** Mark all notifications as read */
export const markAllNotificationsRead = async () => {
  const res = await api.put('/api/v1/notifications/logs/read-all');
  return res.data;
};
